### PR TITLE
[train] Jax Trainer

### PR DESCRIPTION
**Pull request by JiahaoYao**
adding the jax trainer 

- setup the gpu commnication
- adding the `test_gpu` test function

### CODE REVIEW

**Comment by JiahaoYao** [comment]
@Yard1 can u review this?

**Comment by JiahaoYao** [comment]
do i need to add the jax into `requirements` so that ci will install jax?

File: python/ray/train/tests/test_gpu.py [hunk 1]
```diff
@@ -64,6 +67,31 @@ def train_fn():
         )
 
 
+
+ray.init('auto')
```
Comment (Yard1) [inline python/ray/train/tests/test_gpu.py:71]: Remove?

**Review by Yard1 — COMMENTED** [review]
Hey, yeah, you will need to add jax to requirements in`python/requirements/ml/requirements_train.txt`. You also need to run `scripts/format.sh`. I'll make a proper review tomorrow!

File: python/ray/train/tests/test_gpu.py (hunk 1 above)
Comment (JiahaoYao) [reply python/ray/train/tests/test_gpu.py:71]: oh, my bad.

**Comment by JiahaoYao** [comment]
> Hey, yeah, you will need to add jax to requirements in`python/requirements/ml/requirements_train.txt`. You also need to run `scripts/format.sh`. I'll make a proper review tomorrow!

sounds good to me, will add.

File: python/ray/train/examples/jax_mnist_example.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
+import ray.train as train
+from ray.train.trainer import Trainer
+from ray.train.callbacks import JsonLoggerCallback
+
+from typing import Dict
```
Comment (Yard1) [inline python/ray/train/examples/jax_mnist_example.py:5]: could we sort those with `isort`, and keep all ray imports in a second section?

File: python/ray/train/examples/jax_mnist_example.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    """Execute model training"""
+    # get the configuration
+    learning_rate = config['learning_rate']
+    momentum = config['momentum']
+    batch_size = config['batch_size']
+    num_epochs = config['num_epochs']
+    worker_batch_size = batch_size // train.world_size()
+
+    # Create datasets
+    train_ds, test_ds = get_datasets()
```
Comment (Yard1) [inline python/ray/train/examples/jax_mnist_example.py:146]: Is it possible to use ray datasets instead?

File: python/ray/train/jax.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    Args:
+        master_addr_with_port (str): The master work's address plus the port. 
+        num_workers (int): Total number of all the workers.
+        index (int): Index (i.e. world rank) of the current worker.
+    """
+    jax.distributed.initialize(master_addr_with_port, num_workers, index)
+    # TODO
+    # cpu parallel: https://github.com/google/jax/issues/1408
+    # os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count=200"
```
Comment (Yard1) [inline python/ray/train/jax.py:39]: What is this TODO required for? Asking for my own understanding

File: python/ray/train/jax.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+
+
+class JaxBackend(Backend):
+    def on_start(self, worker_group: WorkerGroup, backend_config: JaxConfig):
+        # get the master address
+        master_addr, master_port = worker_group.execute_single(
+            0, get_address_and_port
+        )
+        master_addr_with_port = master_addr + ':' + str(master_port)
```
Comment (Yard1) [inline python/ray/train/jax.py:49]:
```suggestion
        master_addr_with_port = f"{master_addr}:{master_port}"
```

File: python/ray/train/tests/test_gpu.py [hunk 6]
```diff
@@ -64,6 +67,27 @@ def train_fn():
         )
 
 
+def test_jax_get_device(ray_start_4_cpus_2_gpus):
```
Comment (Yard1) [inline python/ray/train/tests/test_gpu.py:70]: Is this test going to work correctly with the number of GPUs required in it? (4 vs 2 initialized)

**Review by Yard1 — COMMENTED** [review]
Hey, thanks, this is looking really good! I have two questions (aside from the comments I made):
1. What is the status of CPU support? We should throw an exception if we don't support it yet.
2. Is it possible to use Ray Datasets in the example? Are there any common preparation steps inside the training function that we can provide utility functions for?
3. We'll need to add documentation for this as well!
4. Should we add more unit tests?

File: python/ray/train/jax.py (hunk 5 above)
Comment (JiahaoYao) [reply python/ray/train/jax.py]: added, thanks!

File: python/ray/train/examples/jax_mnist_example.py (hunk 2 above)
Comment (JiahaoYao) [reply python/ray/train/examples/jax_mnist_example.py:5]: sorted, thanks

File: python/ray/train/jax.py (hunk 4 above)
Comment (JiahaoYao) [reply python/ray/train/jax.py:42]: currently, there is no official support for this. approaches suggested from https://github.com/google/jax/discussions/10745 use MPI4jax.

**Comment by JiahaoYao** [comment]
@Yard1 if we are shifting to ray AIR, do we need the examples in ray train?

**Comment by JiahaoYao** [comment]
this PR needs to merge before https://github.com/ray-project/ray/pull/25385.

File: python/ray/train/examples/jax_mnist_example.py (hunk 3 above)
Comment (JiahaoYao) [reply python/ray/train/examples/jax_mnist_example.py:154]: Yes, see #25385

**Comment by Yard1** [comment]
@JiahaoYao I think it's fine to only have AIR examples.

**Comment by JiahaoYao** [comment]
ready to ship it?

File: python/ray/train/jax.py [hunk 7]
```diff
@@ -10,8 +10,6 @@
 from ray.train.worker_group import WorkerGroup
 from ray.util import PublicAPI
 
-import jax
```
Comment (Yard1) [inline python/ray/train/jax.py]: Let's make sure to add a comment explaining why we don't import jax at the top

File: python/ray/train/examples/jax_mnist_example.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
+import argparse
+import functools
+import time
+from typing import Dict
+
+import einops
+import jax
+import jax.numpy as jnp
```
Comment (Yard1) [inline python/ray/train/examples/jax_mnist_example.py:8]: should jax imports be removed from here?

File: python/ray/train/jax.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    # TODO
+    # cpu parallel: https://github.com/google/jax/issues/1408
+    # os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count=200"
+
+
+def release_tpu_lock():
+    # see https://github.com/google/jax/issues/10192
+    import subprocess
+    subprocess.run("sudo lsof -w /dev/accel0", shell=True)
+    subprocess.run("sudo rm -f /tmp/libtpu_lockfile", shell=True)
```
Comment (Yard1) [inline python/ray/train/jax.py:48]: let's make sure to print a warning (`warnings.warn`) and do this only if `/tmp/libtpu_lockfile` exists

File: python/ray/train/jax.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            setup_futures.append(
+                worker_group.execute_single_async(
+                    i, setup_jax_environment, master_addr_with_port=master_addr_with_port, num_workers=num_workers, index=i
+                )
+            )
+        ray.get(setup_futures)
+
+        # case-insensitivize dict
+        additional_resources_per_worker = worker_group.additional_resources_per_worker
+        additional_resources_per_worker_lower = dict((k.lower(),v) for k,v in additional_resources_per_worker.items())
```
Comment (Yard1) [inline python/ray/train/jax.py:72]:
```suggestion
        additional_resources_per_worker_lower = {k.lower(), v for k, v in additional_resources_per_worker.items()}
```

File: python/ray/train/jax.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        use_tpu = additional_resources_per_worker_lower.pop("tpu", False)
+        # Get setup tasks in order to throw errors on failure.
+        
+        if use_tpu: 
+            # TODO: maybe throw exception or unsafe flag here
```
Comment (Yard1) [inline python/ray/train/jax.py:77]: I think we should throw an exception by default. Users should be able to set an env var to remove the files if they want to explicitly opt into that behavior (we should mention the env var in the exception text)

**Review by Yard1 — COMMENTED** [review]
Thanks, this looks great. Can we make sure to mention in the documentation how the GPU/TPU resources need to be set, and the TPU lock behavior?

I think it's fine to keep the Train example in if it's already written. Is it possible to test it in CI?

Also, we need to make sure CI passes first (for lint - run `scripts/format.sh` locally). Make sure to merge master into this branch too

File: python/ray/train/jax.py (hunk 7 above)
Comment (JiahaoYao) [reply python/ray/train/jax.py]: sgtm

File: python/ray/train/examples/jax_mnist_example.py (hunk 8 above)
Comment (JiahaoYao) [reply python/ray/train/examples/jax_mnist_example.py:8]: no problem

File: python/ray/train/jax.py (hunk 11 above)
Comment (JiahaoYao) [reply python/ray/train/jax.py:77]: added

**Comment by richardliaw** [comment]
Can we instead use AIR instead? Trainer API is going to be deprecated soon.

**Comment by Yard1** [comment]
@richardliaw There is a separate PR that adds the AIR jax trainer. This is required for the backend

**Comment by richardliaw** [comment]
@Yard1 got it

**Comment by JiahaoYao** [comment]
@amogkam test is added now. @Yard1 ready for another quick check.

**Comment by JiahaoYao** [comment]
image
have this issue, https://github.com/JiahaoYao/ray/blob/a26f7a1aaa9c28157f6a1d7eb1726150a45c50d6/python/requirements/ml/requirements_dl.txt#L4-L22

Any suggestion to fix this issue?

File: python/ray/train/jax.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    else:
+        if os.path.isfile("/tmp/libtpu_lockfile"):
+            import warnings
+
+            warnings.warn(
+                "The tpulib lock file is still there, you might"
+                "be unable to use the tpu for training. please "
+                "remove it (`/tmp/libtpu_lockfile`) manually or "
+                "set `RAY_TPU_DEV=1` to release it."
+            )
```
Comment (Yard1) [inline python/ray/train/jax.py:75]:
```suggestion
    tpu_lockfile_path = "/tmp/libtpu_lockfile"
    if RAY_TPU_DEV and os.path.exists(tpu_lockfile_path):
        warnings.warn("Trying to remove the tpulib lock file...")
        subprocess.run("sudo lsof -w /dev/accel0", shell=True)
        subprocess.run(f"sudo rm -f {tpu_lockfile_path}", shell=True)
        if os.path.exists(tpu_lockfile_path):
                warnings.warn(
                    "Couldn't remove the tpulib lockfile "
                    f"(`{tpu_lockfile_path}`)! "
                    "Training will most likely fail or get stuck. "
                    "You may attempt to remove the file manually."
                )
    else:
        if os.path.exists(tpu_lockfile_path):
            warnings.warn(
                "The tpulib lock file is still present. You might"
                "be unable to use the TPU for training. please "
                f"remove it (`{tpu_lockfile_path}`) manually or "
                "set `RAY_TPU_DEV=1` to release it."
            )
```

Move `import warnings` and `import subprocess` to the top please!

File: python/ray/train/jax.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+        # not import jax at the top to avoid tpulib_lockfile error
+        jax.distributed.initialize(master_addr_with_port, num_workers, index)
+
+    # TODO (jimmy)
+    # cpu parallel: https://github.com/google/jax/issues/1408
+    # os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count=200"
+
+
+def release_tpu_lock(RAY_TPU_DEV: bool = False):
```
Comment (Yard1) [inline python/ray/train/jax.py:45]: Maybe we can just do `try_remove_tpulib_lock` instead of `RAY_TPU_DEV` as an argument name here

**Review by Yard1 — APPROVED** [review]
This looks good to me! We should merge the AIR API one ASAP though, and move the tests and example to use the new API.

File: python/requirements/ml/requirements_dl.txt [hunk 14]
```diff
@@ -13,3 +13,9 @@ torchvision==0.10.0;sys_platform=="darwin"
 -f https://download.pytorch.org/whl/torch_stable.html
 torch==1.9.0+cpu;sys_platform!="darwin"
 torchvision==0.10.0+cpu;sys_platform!="darwin"
+
+jax==0.1.69;sys_platform!="darwin"
+jaxlib==0.1.69;sys_platform!="darwin"
+flax==0.4.0;sys_platform!="darwin"
+optax==0.0.9;sys_platform!="darwin"
+einops==0.4.1;sys_platform!="darwin"
```
Comment (ddelange) [inline python/requirements/ml/requirements_dl.txt:21]:
```suggestion
jax==0.1.69
jaxlib==0.1.69
flax==0.4.0
optax==0.0.9
einops==0.4.1
```

to be able to run on CPU Mac?

maybe also with a 'please sync versions' note like on L6?

File: python/requirements_ml_docker.txt [hunk 15]
```diff
@@ -15,3 +15,7 @@ torch-sparse==0.6.12
 # torch-geometric has to be installed after torch-scatter and torch-sparse.
 torch-geometric==2.0.3; python_version < '3.7'
 torch-geometric==2.0.4; python_version >= '3.7'
+
+# Jax
+-f https://storage.googleapis.com/jax-releases/jax_releases.html
+jax[cuda]
```
Comment (ddelange) [inline python/requirements_ml_docker.txt:21]:
```suggestion
jax[cuda]==0.1.69
```

maybe also with a 'please sync versions' note like on L6?

Comment (ddelange) [reply python/requirements_ml_docker.txt:21]: this poses another challenge: the `[cuda]` extra was only [added](https://github.com/google/jax/commit/4d8bce1b8515c68a8432d379d964ea3e6bea463c) in `0.2.23` ([blame](https://github.com/google/jax/blame/a91cb8161340650c2f45f455ab9b5a1178de7658/setup.py#L66)). the `0.1.69` dist does not contain any extras.

py36 support was [dropped](https://github.com/google/jax/commit/94446ff757a3ff1c0dfaba3ecff437ae343c2c71) in `0.2.18`, so another possibility to keep cpu and gpu reqs in sync would be to use [e.g.](https://github.com/google/jax/blob/jax-v0.2.17/setup.py#L19) `jax[cuda111]==0.2.17` here and `jax[cpu]==0.2.17` in `python/requirements/ml/requirements_dl.txt`. that way, you don't need to specify `jaxlib` at all anymore

Comment (JiahaoYao) [reply python/requirements_ml_docker.txt:21]: thanks @ddelange, will take a look soon!

**Comment by JiahaoYao** [comment]
```python
(update_verson) ubuntu@ip-10-0-2-92:~$ pip install optax
Looking in indexes: https://pypi.org/simple, https://pip.repos.neuron.amazonaws.com
Collecting optax
  Using cached optax-0.1.3-py3-none-any.whl (145 kB)
Collecting typing-extensions>=3.10.0
  Using cached typing_extensions-4.3.0-py3-none-any.whl (25 kB)
Requirement already satisfied: numpy>=1.18.0 in ./anaconda3/envs/update_verson/lib/python3.7/site-packages (from optax) (1.19.5)
Collecting jaxlib>=0.1.37
  Downloading jaxlib-0.3.14-cp37-none-manylinux2014_x86_64.whl (71.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 71.3/71.3 MB 11.0 MB/s eta 0:00:00
Requirement already satisfied: jax>=0.1.55 in ./anaconda3/envs/update_verson/lib/python3.7/site-packages (from optax) (0.3.14)
Collecting chex>=0.0.4
  Using cached chex-0.1.3-py3-none-any.whl (72 kB)
Requirement already satisfied: absl-py>=0.7.1 in ./anaconda3/envs/update_verson/lib/python3.7/site-packages (from optax) (0.15.0)
Requirement already satisfied: six in ./anaconda3/envs/update_verson/lib/python3.7/site-packages (from absl-py>=0.7.1->optax) (1.15.0)
Collecting toolz>=0.9.0
  Downloading toolz-0.12.0-py3-none-any.whl (55 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.8/55.8 kB 271.9 kB/s eta 0:00:00
Collecting dm-tree>=0.1.5
  Using cached dm_tree-0.1.7-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (143 kB)
Requirement already satisfied: opt-einsum in ./anaconda3/envs/update_verson/lib/python3.7/site-packages (from jax>=0.1.55->optax) (3.3.0)
Requirement already satisfied: scipy>=1.5 in ./anaconda3/envs/update_verson/lib/python3.7/site-packages (from jax>=0.1.55->optax) (1.7.3)
Requirement already satisfied: etils[epath] in ./anaconda3/envs/update_verson/lib/python3.7/site-packages (from jax>=0.1.55->optax) (0.6.0)
Requirement already satisfied: flatbuffers<3.0,>=1.12 in ./anaconda3/envs/update_verson/lib/python3.7/site-packages (from jaxlib>=0.1.37->optax) (1.12)
Requirement already satisfied: zipp in ./anaconda3/envs/update_verson/lib/python3.7/site-packages (from etils[epath]->jax>=0.1.55->optax) (3.8.1)
Requirement already satisfied: importlib_resources in ./anaconda3/envs/update_verson/lib/python3.7/site-packages (from etils[epath]->jax>=0.1.55->optax) (5.8.0)
Installing collected packages: dm-tree, typing-extensions, toolz, jaxlib, chex, optax
  Attempting uninstall: typing-extensions
    Found existing installation: typing-extensions 3.7.4.3
    Uninstalling typing-extensions-3.7.4.3:
      Successfully uninstalled typing-extensions-3.7.4.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.6.2 requires typing-extensions~=3.7.4, but you have typing-extensions 4.3.0 which is incompatible.
Successfully installed chex-0.1.3 dm-tree-0.1.7 jaxlib-0.3.14 optax-0.1.3 toolz-0.12.0 typing-extensions-4.3.0
```

image



i see @amogkam , the issue comes from installing the `optax` and then causing some issues from `typing-extensions`.

**Comment by JiahaoYao** [comment]
this error looks like the error shown [here](https://github.com/tensorflow/tensorflow/issues/51743)

**Comment by JiahaoYao** [comment]
@ddelange really appreciate the verison suggestion!

**Comment by JiahaoYao** [comment]
sorry for the late replay, but the version suggestion is so precious! i was blocked at the dependencies! now, i can ship the jax stuff.

**Comment by ddelange** [comment]
will probably need rebase after https://github.com/ray-project/ray/pull/28808

**Comment by JiahaoYao** [comment]
keep alive

**Comment by ddelange** [comment]
fyi #26072 was merged

**Pull request closed without merging**